src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1577,6 +1595,66 @@ public override BoundNode VisitCall(BoundCall node)
             return null;
         }
 
+        private bool MethodReturnsTrueWhenNotNull(MethodSymbol method)
+        {
+            if (method.ReturnType.SpecialType != SpecialType.System_Boolean)
+            {
+                return false;
+            }
+
+            return method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_String_IsNullOrEmpty));
+        }
+
+        private void VisitTrueWhenNotNull(BoundExpression operand, TypeSymbol boolType)
+        {
+            if (!operand.Type.IsReferenceType)
+            {
+                return;
+            }
+
+            // operand == null || someUnknownBoolean
+            BoundExpression equivalent = LogicalOr(
+                new BoundBinaryOperator(operand.Syntax, BinaryOperatorKind.Equal, operand,
+                    new BoundLiteral(operand.Syntax, ConstantValue.Null, operand.Type),
+                    constantValueOpt: null, methodOpt: null, LookupResultKind.Viable, boolType),
+                new BoundLiteral(operand.Syntax, constantValueOpt: null, boolType));
```

Comment:
We're tracking "maybe null" rather than "definitely null". In short, is the `|| someUnknownBoolean` necessary?

Reply:
Can this be implemented without synthesizing nodes?